torch/csrc/deploy/test_deploy.cpp

```diff
@@ -0,0 +1,126 @@
+#include <gtest/gtest.h>
+#include <torch/deploy.h>
+#include <torch/script.h>
+#include <torch/torch.h>
+#include <future>
+#include <iostream>
+#include <string>
+
+int main(int argc, char* argv[]) {
+  ::testing::InitGoogleTest(&argc, argv);
+  int rc = RUN_ALL_TESTS();
+  return rc;
+}
+
+void compare_torchpy_jit(const char* model_filename, const char* jit_filename) {
+  // Test
+  torch::InterpreterManager m(1);
+  torch::Package p = m.load_package(model_filename);
+  auto model = p.load_pickle("model", "model.pkl");
+  at::IValue eg;
+  {
+    auto I = p.acquire_session();
+    eg = I.self.attr("load_pickle")({"model", "example.pkl"}).toIValue();
+  }
+
+  at::Tensor output = model(eg.toTuple()->elements()).toTensor();
+
+  // Reference
+  auto ref_model = torch::jit::load(jit_filename);
+  at::Tensor ref_output =
+      ref_model.forward(eg.toTuple()->elements()).toTensor();
+
+  ASSERT_TRUE(ref_output.allclose(output, 1e-03, 1e-05));
+}
+
+const char* simple = "torch/csrc/deploy/example/generated/simple";
+const char* simple_jit = "torch/csrc/deploy/example/generated/simple_jit";
+
+const char* path(const char* envname, const char* path) {
+  const char* e = getenv(envname);
+  return e ? e : path;
+}
+
+TEST(TorchpyTest, SimpleModel) {
+  compare_torchpy_jit(path("SIMPLE", simple), path("SIMPLE_JIT", simple_jit));
+}
+
+TEST(TorchpyTest, ResNet) {
+  compare_torchpy_jit(
+      path("RESNET", "torch/csrc/deploy/example/generated/resnet"),
+      path("RESNET_JIT", "torch/csrc/deploy/example/generated/resnet_jit"));
+}
+
+TEST(TorchpyTest, Movable) {
+  torch::InterpreterManager m(1);
+  torch::MovableObject obj;
+  {
+    auto I = m.acquire_one();
+    auto model =
+        I.global("torch.nn", "Module")(std::vector<torch::PythonObject>());
+    obj = I.create_movable(model);
+  }
+  obj.acquire_session();
+}
+
+TEST(TorchpyTest, MultiSerialSimpleModel) {
+  torch::InterpreterManager manager(3);
+  torch::Package p = manager.load_package(path("SIMPLE", simple));
+  auto model = p.load_pickle("model", "model.pkl");
+  auto ref_model = torch::jit::load(path("SIMPLE_JIT", simple_jit));
+
+  auto input = torch::ones({10, 20});
+  size_t ninterp = 3;
+  std::vector<at::Tensor> outputs;
+
+  // Futures on model forward
+  for (size_t i = 0; i < ninterp; i++) {
+    outputs.push_back(model({input}).toTensor());
+  }
+
+  // Generate reference
+  auto ref_output = ref_model.forward({input}).toTensor();
+
+  // Compare all to reference
+  for (size_t i = 0; i < ninterp; i++) {
+    ASSERT_TRUE(ref_output.equal(outputs[i]));
+  }
+}
+
+TEST(TorchpyTest, ThreadedSimpleModel) {
```

Comment:
not that it would affect actual coverage, but it might be good to include tests of any other expected/popular threading model such as in your paper the OMP for